pkg/cmd/taskrun/logs.go

```diff
@@ -0,0 +1,110 @@
+package taskrun
+
+import (
+	"fmt"
+	"github.com/tektoncd/cli/pkg/cli"
+	"github.com/tektoncd/cli/pkg/logs"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	"github.com/tektoncd/pipeline/pkg/reconciler/v1alpha1/taskrun/resources"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type TaskRunLogs struct {
+	TaskName string
+	Run      string
+	Ns       string
+	Status   v1alpha1.TaskRunStatus
+	Clients  *cli.Clients
+}
+
+type LogOptions struct {
+	Follow   bool
+	AllSteps bool
+}
+
+func (tr TaskRunLogs) Fetch(flags LogOptions, stream logs.Streams, reader *logs.LogFetcher) {
+	var (
+		taskPod  = tr.Status.PodName
+		taskName = tr.TaskName
+
+		kube = tr.Clients.Kube
+	)
+
+	if !taskRunHasStarted(tr.Status) {
+		fmt.Fprintf(stream.Out, "Task %s has not started yet \n", taskName)
+		return
+	}
+
+	pod, err := kube.CoreV1().
+		Pods(tr.Ns).
+		Get(taskPod, v1.GetOptions{})
+	if err != nil {
+		fmt.Fprintf(stream.Err, "Error in getting pod: %s \n", err)
+		return
+	}
+
+	if !podHasStarted(pod) {
+		fmt.Fprintf(stream.Out, "Task %s pod %s has not started yet \n", taskName, taskPod)
+		return
+	}
+
+	steps := filterSteps(tr.Status, pod, flags.AllSteps)
+
+	for _, step := range steps {
+		if !stepHasStarted(step) {
+			fmt.Fprintf(stream.Out, "Step %s has not started yet \n", taskName)
+			continue
+		}
+
+		pl := logs.NewPodLogs(taskPod, tr.Ns, reader)
+		err := pl.Fetch(stream, ContainerNameForStep(step.Name), flags.Follow, func(s string) string {
+			return fmt.Sprintf("[%s : %s] %s", taskName, step.Name, s)
+		})
+
+		if err != nil {
+			fmt.Fprintf(stream.Err, "Error in printing logs for the %s : %s \n", step.Name, err)
+		}
+
+		fmt.Fprint(stream.Out, "\n")
+	}
+}
+
+func podHasStarted(pod *corev1.Pod) bool {
+	return !(pod.Status.Phase == corev1.PodPending || pod.Status.Phase == corev1.PodUnknown)
+}
+
+func taskRunHasStarted(trStatus v1alpha1.TaskRunStatus) bool {
+	return trStatus.StartTime != nil && !trStatus.StartTime.IsZero()
+}
+func filterSteps(trStatus v1alpha1.TaskRunStatus, pod *corev1.Pod, allSteps bool) []v1alpha1.StepState {
+	if !allSteps {
+		return trStatus.Steps
+	}
+
+	initSteps := []v1alpha1.StepState{}
+	for _, ics := range pod.Status.InitContainerStatuses {
+		initSteps = append(initSteps, v1alpha1.StepState{
+			ContainerState: *ics.State.DeepCopy(),
+			Name:           resources.TrimContainerNamePrefix(ics.Name),
+		})
+	}
+	//append normal steps to preserve the order
+	initSteps = append(initSteps, trStatus.Steps...)
+
+	return initSteps
+}
+
+func stepHasStarted(stepState v1alpha1.StepState) bool {
+	return stepState.ContainerState.Waiting == nil
+}
+
+//TODO: annonymous steps?
+func ContainerNameForStep(name string) string {
+	switch name {
+	case "nop":
+		return "nop"
+	default:
+		return "build-step-" + name
```

Comment:
I'm not sure if we need to prepend the `step` names with `build-step`, it feels weird :sweat_smile:

Reply:
yeah, that's how container name as are formed here. we need container name at the end to query the logs.